### PR TITLE
issues 98 and 99

### PR DESCRIPTION
**Pull request by JinoBetti**

### CODE REVIEW

**Review by Ooouch1 — CHANGES_REQUESTED** [review]
Build by maven fails since the test of `OriLine::hashCode` fails.

**Comment by Ooouch1** [comment]
[heavy_test.zip](https://github.com/oripa/oripa/files/6371824/heavy_test.zip)
A .opx file of 100x100 grid mountain lines with diagonals for performance test.

File: src/main/java/oripa/domain/cptool/LineAdder.java [hunk 1]
```diff
@@ -204,58 +189,76 @@ public void addLine(final OriLine inputLine, final Collection<OriLine> currentLi
...
+		overlappingLines.forEach(currentLine -> {
+				List<OriLine> oriLines = splitOverlappingLines(currentLine, splitNewLine);
+				currentLines.remove(currentLine);
+				currentLines.addAll(oriLines);
+			});
+	}
+
+	private List<OriLine> getOverlappingLines(Collection<OriLine> currentLines, OriLine splitNewLine) {
+		return currentLines.stream()
+				.filter(oriLine -> oriLine.getType() != OriLine.Type.CUT)
```
[the remark this answers is not in the document]
Comment (JinoBetti) [reply src/main/java/oripa/domain/cptool/LineAdder.java]: I changed it before seeing your comment actually, as soon as I saw a method already existed for that 👍 great review tip!

File: src/main/java/oripa/value/OriLine.java [hunk 2]
```diff
@@ -178,33 +208,25 @@ public int compareTo(final OriLine oline) {
...
-		int comparison10 = this.p1.compareTo(oline.p0);
-		int comparison11 = this.p1.compareTo(oline.p1);
+		OriLine that = (OriLine) obj;
 
 		// same direction?
+		int comparison00 = this.p0.compareTo(that.p0);
+		int comparison11 = this.p1.compareTo(that.p1);
 		if (comparison00 == 0 && comparison11 == 0) {
-			return true;
+			return this.type.equals(that.type);
```
[the remark this answers is not in the document]
Comment (JinoBetti) [reply src/main/java/oripa/value/OriLine.java:222]: fixed, and I also had to fix the mocks from OutlineAdderTest to change the line type for the eq() ArgumentMatcher

File: src/main/java/oripa/value/OriLine.java [hunk 3]
```diff
@@ -33,6 +38,20 @@
 	private static final int TYPE_VALLEY = 3;
 	private static final int TYPE_CUT_MODEL = 4;
 
+	public boolean isVertical() {
+		return abs(p0.x - p1.x) < EPS;
+	}
+
+	public boolean contains(OriPoint oriPoint) {
+		RectangleClipper rectangleClipper = new RectangleClipper(p0.x, p0.y, p1.x, p1.y);
+		if(isVertical()) return abs(getAffineXValueAt(oriPoint.y) - oriPoint.x) < EPS && rectangleClipper.contains(oriPoint);
```
[the remark this answers is not in the document]
Comment (JinoBetti) [reply src/main/java/oripa/value/OriLine.java]: Agreed, and fixed 👍

**Comment by JinoBetti** [comment]
> [heavy_test.zip](https://github.com/oripa/oripa/files/6371950/heavy_test.zip)
> A .opx file of 100x100 grid mountain lines with diagonals for performance test.
> 
> edit: the file was incomplete. I re-uploaded.

wow, that's a lot of lines... do you have the standards of how long it takes with the current version, and how long you find it "acceptable"?, it's long to load but not suprising (~2-3 seconds)... cutting and pasting the whole thing is very long, though (47 seconds), but again... nothing very surprising for something that appears as a big black square due to the number of vertices :)

**Review by Ooouch1 — CHANGES_REQUESTED** [review]
I got a wrong result:

1. Draw (-100,0, 0,0, V) and (100,0, 200,0, V).
2. Draw (-200,0, 200,0, M).
3. You will get some extra mountain lines. Try deleting the generated lines one by one.

**Comment by Ooouch1** [comment]
> do you have the standards of how long it takes with the current version,
> and how long you find it "acceptable"?, it's long to load but not suprising (~2-3 seconds)... cutting and pasting the whole thing is very long, though (47 seconds)

On ver. 1.45, if I cut all lines of heavy_test.opx and paste them, it takes 1~2 seconds with Surface Pro 7 core i5.
Since `OverlappingLineExtractor::extract(lines)` takes around 1 second in this case (you can try this method by running foldability check), I guess there is a way to finish detecting overlapping segments and splitting them in less than 5 seconds on my Surface.
Maybe a faster algorithm is to use a modification of `OverlappingLineExtractor::extract(lines)` to group the segments with the support lines.

By the way, it was good if you made pull requests per issue, a PR for issue 98 and a PR for issue 99, which enables me to merge fix of issue 98 earlier.

**Comment by JinoBetti** [comment]
it takes ~8 seconds with the old version on my desktop. The overlap detection of each of the 30 000 input lines (squared...) is indeed much too long in comparison :( [900 000 000 iterations in 37484ms] 
I'll see in the next few days if I can find a better way to do it, probably using your idea of grouping by support line. 

The only other solution I can think of is having a reduced functionality mode when adding too many lines at once (When pasting more than 250 lines at once, fall back to the old way for example, since the performance problem is only visible when we have a huge number of input lines and a huge number of existing lines [e.g. : 7 500 000 iterations in 479ms], and I never used than many lines before so I think it wont happen very often)

Anyways, I still have work to do to do it correctly, but I'm sure we'll find a way... ganbarimasu!

**Comment by lifesbest23** [comment]
> I got a wrong result:
> 
>     1. Draw (-100,0, 0,0, V) and (100,0, 200,0, V).
> 
>     2. Draw (-200,0, 200,0, M).
> 
>     3. You will get some extra mountain lines. Try deleting the generated lines one by one.

I fixed this issue as well as changed the hashing Function to include `type` here: https://github.com/lifesbest23/oripa/tree/fix/issue98

> the programmer should be aware that producing distinct integer results for unequal objects may improve the performance of hash tables.
> As much as is reasonably practical, the `hashCode `method defined by class `Object` does return distinct integers for distinct objects.

I am not sure if my implementation is better, since it has more distinct hashes for `OriLine`, but it does not improve anything performance wise in my testing, so I am not sure if the previous implementation is preferred.

> On ver. 1.45, if I cut all lines of heavy_test.opx and paste them, it takes 1~2 seconds with Surface Pro 7 core i5.
> Since `OverlappingLineExtractor::extract(lines)` takes around 1 second in this case

The cut paste heavy_test does seem to work as well as in the current master branch after the changes in this PR , at least for me, so I think this is a separate issue.

I am not sure where to make a PR though, either for @JinoBetti s branch, or a new one here?

**Comment by Ooouch1** [comment]
@lifesbest23 It seems you have solved the problem of overwriting line type by using "1 new overlapping line vs. n existing lines" case solution. It might be slow if many added lines overlap with existing lines. Is the performance on heavy_test.opx OK if you change the left half of the creases to V and the right half to M, and cut/copy the V lines then paste over M lines?

Your implementation of `OriLine::hashCode` seems to violate the 2nd [condition](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#hashCode--) of `hashCode()`:
*If two objects are equal according to the equals(Object) method, then calling the hashCode method on each of the two objects must produce the same integer result.*

> I am not sure where to make a PR though, either for @JinoBetti s branch, or a new one here?

I think new PR on this repository will make reviews between you and me easier.

**Pull request merged**